Incremental commits spec changes (#1511)
input-output-hk/hydra: spec/onchain.tex

	\item one output paying to $\nuHead$ with value $\valHead'$ and
		  datum $\datumHead'$,
\end{itemize}

\noindent The state-machine validator $\nuHead$ is spent with
$\redeemerHead = (\mathsf{increment}, \xi, \txOutRef_{\mathsf{increment}})$, where $\xi$ is a multi-signature of
the increment snapshot which authorizes addition of some UTxO $U_\alpha$ and $\txOutRef_{\mathsf{increment}}$ are the output references we need to spend. The
validator checks:
\begin{menumerate}
	\item State is advanced from $\datumHead \sim \stOpen$ to
	$\datumHead' \sim \stOpen$, parameters $\cid,\hydraKeysAgg,\nop,\cPer$
	stay unchanged and the new state is governed again by $\nuHead$:
	\[
		(\stOpen,\cid,\hydraKeysAgg,\nop,\cPer,v,\eta) \xrightarrow[\xi, \txOutRef_{\mathsf{increment}}]{\mathsf{increment}} (\stOpen,\cid,\hydraKeysAgg,\nop,\cPer,v',\eta')
	\]
	\item New version $v'$ is incremented correctly
	\[
		v' = v + 1
	\]
\item $\xi$ is a valid multi-signature of the new snapshot state (currency id $\cid$, the current state version $v$, the new state $\eta'$, UTxO to deposit $\eta_\alpha$)
	\[
    \msVfy(\hydraKeys,(\cid || v || s || \eta' || \eta_\alpha || \bot),\xi) = \true
	\]

>   where snapshot number $s$ is provided through the redeemer, $\eta_\alpha$ is the digest of all added UTxO.

Reviewer: Must: add `s` to the redeemer definition $\rho_\mathsf{head}$ above

	\[

> 		\eta_\alpha = \hash(\bigoplus_{j=1}^m \bytes(i_{j}))

Reviewer: ```suggestion
		\eta_\alpha = \hash(\bigoplus_{j=1}^m \bytes(o_\mathsf{increment}_j}))
```

		$\stOpen$ head output.}\label{fig:collectComTx}
\end{figure}

\subsection{\red{Increment Transaction}}\label{sec:increment-tx}

\noindent The \mtxIncrement{} transaction (Figure~\ref{fig:incrementTx}) allows a party to
add a UTxO to an already open head and consists of:

\begin{itemize}
	\item one input spending from $\nuHead$ holding the $\st$ with $\datumHead$,
	\item one output paying to $\nuHead$ with value $\valHead'$ and
		  datum $\datumHead'$,

>     \item one or more inputs spent in this transaction $i_{1} \dots i_m$ where $m$ is the number of inputs to include and $\forall i \in \{i_{1} \dots i_{m}\}, i \notin I_{head}$.

Reviewer: What is $I_{head}$?

\end{itemize}

\noindent The state-machine validator $\nuHead$ is spent with

> $\redeemerHead = (\mathsf{increment}, \xi)$, where $\xi$ is a multi-signature of

Reviewer: Must: include a list of output references (inputs) in the redeemer to resolve the UTxO we want to consider (the question mark we put on the transaction traces + red annotation how to work around that)

the increment snapshot which authorizes addition of some UTxO $U_\alpha$. The
validator checks:
\begin{menumerate}
	\item State is advanced from $\datumHead \sim \stOpen$ to
	$\datumHead' \sim \stOpen$, parameters $\cid,\hydraKeysAgg,\nop,\cPer$
	stay unchanged and the new state is governed again by $\nuHead$:
	\[
		(\stOpen,\cid,\hydraKeysAgg,\nop,\cPer,v,\eta) \xrightarrow[\xi]{\mathsf{increment}} (\stOpen,\cid,\hydraKeysAgg,\nop,\cPer,v',\eta')
	\]
	\item Increment snapshot number $s'$ is higher than the currently stored snapshot number $s$
	\[
		s' > s

> 	\]

Reviewer: Must: remove the snapshot number check (as also in the decrement)

	\item New version $v'$ is incremented correctly
	\[
		v' = v + 1
	\]
	\item $\xi$ is a valid multi-signature of the new snapshot state (currency id $\cid$, the current state version $v$, the new state $\eta'$, and $\eta_\theta$)
	\[

>     \msVfy(\hydraKeys,(\cid || v || s || \eta' || \eta_\theta || \red{\theta}),\xi) = \true

Reviewer: What is $\theta$? (not explained up to this point)

	\]
  where snapshot number $s$ is provided through the redeemer, and $\eta_\theta$ is the digest of all added UTxO
	\[

> 		\eta_\theta = U^{\#}_\theta  = \hash(\bigoplus_{j=1}^m \bytes(i_{j}))

Reviewer: This is not so easy, transaction inputs $i_j$ are just "out refs" and to get the right outputs, we'd need to look them up. Also, it's no where mentioned _which_ inputs are to be considered.

Use a similar approach as in the commit section. That is, add a list of output references $\underline{\phi}_{\mathsf{committed}}$ to the redeemer and mention in the intro that "zero or more inputs with reference $\phi_j$ spending output $o_j$ with $\mathsf{val}_j$". That way, you can later reference the corresponding outputs here using $o_j$

>   where snapshot number $s$ is provided through the redeemer, $\eta_\alpha$ is the digest of all added UTxO, and $\eta_\omega$ is the digest of all removed UTxO but in case of increment transaction $\eta_\omega = \bot$.

Reviewer: Should: just put $\bot$ directly into the expression above. No need to explain `$\eta_\omega$ here.

\noindent The state-machine validator $\nuHead$ is spent with
$\redeemerHead = (\mathsf{decrement}, \xi, s, m)$, where $\xi$ is a multi-signature of

> the decrement snapshot which authorizes removal of some UTxO $\red{U_\theta}$, $s$ is the

Reviewer: Actually, we do not need to give this UTxO a variable here (we don't use it in the constraints).

```suggestion
the decrement snapshot which authorizes removal of some UTxO, $s$ is the
```

	\[
		(\stOpen,\cid,\hydraKeys,\cPer,v,\eta) \xrightarrow[\xi, s, m]{\mathsf{decrement}} (\stOpen,\cid,\hydraKeys,\cPer,v',\eta')
	\]
	\item Decrement snapshot number $s'$ is higher than the currently stored snapshot number $s$
	\[
		s' > s

> 	\]

Reviewer: Must: not include this check (there is no snapshot number in the open state to compare)

	\item New version $v'$ is incremented correctly
	\[
		v' = v + 1
	\]

> 	\item $\xi$ is a valid multi-signature of the new snapshot state (currency id $\cid$, the current state version $v$, the new state $\eta'$, and $\red{\eta_\theta}$)

Reviewer: The removed UTxO $\eta$ is defined below, no need to mention it here.

```suggestion
	\item $\xi$ is a valid multi-signature of the new snapshot state (currency id $\cid$, the current state version $v$, the new state $\eta'$)
```

	\[

> 		\msVfy(\hydraKeys,(\cid || v || s || \eta' || \red{\eta_\theta} || \red{\theta}),\xi) = \true

Reviewer: Must: define what $\theta$ is and where it comes from

	\]

>     where snapshot number $s$ is provided through the redeemer \red{, $\eta_\alpha$ is a digest of all added UTxO (in case of decrement transaction $\eta_\alpha = \bot$)} and $\eta_\omega$ must be the digest of all removed UTxO.

Reviewer: Should: Just put $\bot$ in the expression and don't use $\eta_\alpha$ (simplifies definition IMO)

	$\datumHead' \sim \stClosed$, parameters $\cid,\hydraKeys,\cPer$
	stay unchanged and the new state is governed again by $\nuHead$
	\[

>     (\stOpen,\cid,\hydraKeys,\cPer,v,\eta) \xrightarrow[\mathsf{CloseType}]{\stClose} (\stClosed,\cid,\hydraKeys,\cPer,v',s',\eta', \eta_\Delta', \contesters,\Tfinal)

Reviewer: Could: Close figure is outdated, could make it consistent again.

		\[
			\eta' = \eta
		\]
    \red{
		  \item $\mathsf{UnusedInc}$: Closing snapshot refers to current state version $v$ and any UTxO to increment must not be recorded in the closed state.
		  \[

> 		  	\eta_\alpha = \eta_\Delta'

Reviewer: Fixed wrongly, it should read this instead

```suggestion
		  	\eta_\Delta' = \bot
```

		\[
			\eta' = \eta
		\]
    \red{

> 		  \item $\mathsf{UnusedInc}$: Closing snapshot refers to current state version $v$ and any UTxO to increment must be $\bot$ in the closed state.

Reviewer: ```suggestion
		  \item $\mathsf{UnusedInc}$: Closing snapshot refers to current state version $v$ and any UTxO to increment of the closing snapshot must not be recorded in the closed state.
```

		  \[

> 		  	\eta_\alpha = \eta_\Delta' = \bot

Reviewer: This is wrong: $\eta_\alpha$ might be $\neq \bot$!

```suggestion
		  	\eta_\Delta' = \bot
```

Author: How come? In the `UnusedInc` we provide $\eta_\alpha$ in the redeemer but it should be $\bot$ in the closed state (which is $\eta_\Delta$ here)

Reviewer: As it's a snapshot that approves an increment, it naturally would include a UTxO to commit. Consequently $\eta_\alpha$ of the snapshot signature is $\eta_\alpha \neq \bot$

Author: ok we are on the same page, only $\eta_\Delta$ is $\bot$ in the closed state

		  \]
		  \[
		  	\eta_\omega = \bot
		  \]
		  \[
		  \msVfy(\hydraKeys,(\cid || v || s' || \eta' || \eta_\alpha || \eta_\omega),\xi) = \true
		  \]
      where $\eta_\alpha$ is provided by the redeemer \footnote{$\eta_\alpha$ needs to be provided to verify the signature, but is otherwise not relevant for the closed state}.
    }
		\item $\mathsf{UnusedDec}$: Closing snapshot refers to current state version $v$ and any UTxO to decrement need to be present in the closed state too.
		\[
			\eta_\omega = \eta_\Delta' \neq \bot
		\]
    \red{
		  \[
		  	\eta_\alpha = \bot
		  \]
    }
		\[
      \msVfy(\hydraKeys,(\cid || v || s' || \eta' \red{|| \eta_\alpha} || \eta_\omega),\xi) = \true
		\]

>     which implies $\eta_\Delta' = \eta_\omega$.
> 		% TODO: put more into the redeemer to simplify?

Reviewer: ```suggestion
```

		\todo{this is hard to understand}
    \red{

> 		  \item $\mathsf{UsedInc}$: Closing snapshot refers the previous state $v - 1$ and any UTxO to increment of the closing snapshot must be recorded in the closed state.

Reviewer: ```suggestion
		  \item $\mathsf{UsedInc}$: Closing snapshot refers the previous state $v - 1$ and any UTxO to increment must be recorded in the closed state.
```

		  \[
		  	\eta_\alpha = \eta_\Delta' \neq \bot
		  \]
		  \[
		  	\eta_\omega = \bot
		  \]
		  \[
		  	\msVfy(\hydraKeys,(\cid || v - 1 || s' || \eta' || \eta_\alpha || \eta_\omega ),\xi) = \true
		  \]

>       which implies $\eta_\Delta' = \eta_\alpha$.

Reviewer: ```suggestion
```

		\[
			\eta' = \eta
		\]
    \red{
		  \item $\mathsf{UnusedInc}$: Closing snapshot refers to current state version $v$ and any UTxO to increment must be $\bot$ in the closed state.
		  \[

> 		  	\eta_\alpha' = \eta_\alpha = \eta_\Delta' = \bot

Reviewer: What is $\eta_\alpha'$ here?

Note that $=$ is **not** an assignment here, but an equality. So if you use variables, they need to refer to something.

I don't think that we need the $\eta_\alpha'$ and $\eta_\omega'$ in any of these cases

    }
		\item $\mathsf{UnusedDec}$: Closing snapshot refers to current state version $v$ and any UTxO to decrement need to be present in the closed state too.
		\[
			\eta_\omega' = \eta_\omega = \eta_\Delta' \neq \bot
		\]
    \red{
		  \[
		  	\eta_\alpha' = \bot
		  \]
    }
		\[
      \msVfy(\hydraKeys,(\cid || v || s' || \eta' \red{|| \eta_\alpha'} || \eta_\omega'),\xi) = \true
		\]
    which implies $\eta_\Delta' = \eta_\omega$.
		% TODO: put more into the redeemer to simplify?
		\todo{this is hard to understand}
    \red{
		  \item $\mathsf{UsedInc}$: Closing snapshot refers the previous state $v - 1$ and any UTxO to increment of the closing snapshot must be recorded in the closed state.
		  \[
		  	\eta_\alpha' = \eta_\alpha = \eta_\Delta' \neq \bot
		  \]
		  \[
		  	\eta_\omega' = \bot
		  \]
		  \[
		  	\msVfy(\hydraKeys,(\cid || v - 1 || s' || \eta' || \eta_\alpha' || \eta_\omega' ),\xi) = \true
		  \]
      which implies $\eta_\Delta' = \eta_\alpha$.
    }
		\item $\mathsf{UsedDec}$: Closing snapshot refers the previous state $v - 1$ and any UTxO to decrement of the closing snapshot must not be recorded in the closed state.
		\[

> 	 		\eta_\omega' = \eta_\omega = \eta_\Delta' \neq \bot

Reviewer: This is wrong!

In the diff you can clearly see that we explicitly required $\eta_{\Delta}' = \bot$ before and this is still the case (the english sentence above also claims that).

		\[
			\eta' = \eta
		\]
		\red{
			\[
			\theta = Normal

> 			\]

Reviewer: This check is pointless: $\theta$ is not linked with anything on-chain and can be freely picked in the redeemer.

			\item $\mathsf{Unused/Inc}$: Closing snapshot refers to current state version $v$, snapshot type $\theta = Inc$ and any UTxO to increment must be $\bot$ in the closed state.

> 			\[
> 			\msVfy(\hydraKeys,(\cid || v || s' || \eta' || \eta_{\Delta}' || \theta),\xi) = \true
> 			\]
> 			which implies $\eta_{\Delta}' = \eta_\theta$ while $\eta_\theta$ must not be provided.

Reviewer: This is incorrect! The sentence in 483 is correct, but the expression here says that $\eta_\theta$ should end up in the closed state!

```suggestion
                        \[
                          \eta_{\Delta}' = \bot
                        \]
			\[
			  \msVfy(\hydraKeys,(\cid || v || s' || \eta' || \eta_\theta || \theta),\xi) = \true
			\]
			where $\eta_\theta$ is provided by the redeemer\footnote{$\eta_\theta$ needs to be provided to verify the signature, but is otherwise not relevant for the closed state}.
```

			\item $\mathsf{Unused/Dec}$: Closing snapshot refers to current state version $v$, snapshot type $\theta = Dec$ and any UTxO to decrement need to be present in the closed state too.
			\[
        \msVfy(\hydraKeys,(\cid || v || s' || \eta' || \eta_{\Delta}' || \theta),\xi) = \true
			\]

> 			which implies $\eta_{\Delta}' = \eta_\theta$ while $\eta_\theta$ must be provided.

Reviewer: ```suggestion
			which implies $\eta_{\Delta}' = \eta_\theta$ while $\eta_\theta$ does not need to be provided separately.
```

			% TODO: put more into the redeemer to simplify?
			\todo{this is hard to understand}
			\item $\mathsf{Used/Dec}$: Closing snapshot refers the previous state $v - 1$ , snapshot type $\theta = Dec$  and any UTxO to decrement of the closing snapshot must not be recorded in the closed state.
			\[
				\eta_{\Delta}' = \bot
			\]
			\[
        \msVfy(\hydraKeys,(\cid || v - 1 || s' || \eta' || \eta_\theta || \theta),\xi) = \true
			\]
			where $\eta_\theta$ is provided by the redeemer\footnote{$\eta_\theta$ needs to be provided to verify the signature, but is otherwise not relevant for the closed state}.
			\item $\mathsf{Used/Inc}$: Closing snapshot refers the previous state $v - 1$ , snapshot type $\theta = Inc$  and any UTxO to increment of the closing snapshot must be recorded in the closed state.

> 			\[
> 				\eta_{\Delta}' \neq \bot
> 			\]
> 			\[
> 				\msVfy(\hydraKeys,(\cid || v - 1 || s' || \eta' || \eta_\theta || \theta),\xi) = \true
> 			\]
> 			where $\eta_\theta$ must be provided in the closed state.

Reviewer: No tonly $\neq \bot$, but $= \eta_\theta$!

```suggestion
			\[
				\eta_{\Delta}' = \eta_\theta
			\]
			\[
				\msVfy(\hydraKeys,(\cid || v - 1 || s' || \eta' || \eta_\theta || \theta),\xi) = \true
			\]
```

This is, in fact, an alternative way to describe the same case as in (c). Implementation wise, the equality $\eta_{\Delta}' = \eta_\theta$ allows us to insert $\eta_{\Delta}'$ directly into $(\mathsf{cid} || v - 1 || s' || \eta_{\Delta}' || \eta_\theta || \theta),\xi)$.. which gives the above definition in (c).

Pick one way of describing it.

		\[
      \red{\eta_\Delta' \neq \bot}
		\]
		\[
      \msVfy(\hydraKeys,(\cid || v || s' || \eta' || \eta_\Delta || \red{\theta}),\xi) = \true
		\]
		which implies $\eta_\Delta' = \eta'_\theta$ while $\eta'_\theta$ does not need to be provided.
    \item $\mathsf{Used/Inc}$: Contesting snapshot refers the previous state version $v - 1$ \red{, snapshot type $\theta = Inc$} and any UTxO to \red{increment} must be recorded in the closed state.
		\[
      \red{\eta_\Delta' = \bot}
		\]
		\[
      \msVfy(\hydraKeys,(\cid || v - 1 || s' || \eta' || \red{\eta'_\theta} || \red{\theta}),\xi) = \true
		\]
		which implies $\eta_\Delta' = \eta'_\theta$ while $\eta'_\theta$ does not need to be provided.
    \item $\mathsf{Used/Dec}$: Contesting snapshot refers the previous state version $v - 1 $\red{, snapshot type $\theta = Dec$} and any UTxO to decrement must not be recorded in the closed state.
		\[
			\eta_\Delta' = \bot
		\]
		\[
      \msVfy(\hydraKeys,(\cid || v - 1 || s' || \eta' || \red{\eta'_\theta} || \red{\theta}),\xi) = \true
		\]

> 		where $\eta_\theta$ is provided by the redeemer.

Reviewer: Not reviewed, but must follow changes from closeTx directly.

	\[
		s' > s
	\]

> \item $\mathsf{ContestType} = \red{(\mathsf{UnusedInc}, \xi, \eta_\alpha)} \cup (\mathsf{UnusedDec}, \xi) \red{\cup (\mathsf{UsedInc}, \xi)}  \cup (\mathsf{UsedDec}, \xi, \eta_\omega)$, where $\xi$ is a multi-signature of the contesting snapshot \red{and $\eta_\alpha$} and $\eta_\omega$ are the digests of the UTxO to \red{increment or} decrement respectively, \red{four} cases are distinguished:

Reviewer: See all comments from close above and make sure they are fixed here to (if they apply)
